### PR TITLE
Make workflow show output json usable for replay

### PR DESCRIPTION
**Pull request by feedmeapples**
## What was changed

Made `workflow show --output json` produce replayable history

Depends on https://github.com/temporalio/tctl-kit/pull/39

## Why?

Allows exporting replayable history with syntax

```
workflow show --output json -w workflow_id > ~/history.json
```

## Checklist

1. Closes 

2. How was this tested:

E2E test here https://github.com/temporalio/cli/pull/171

3. Any docs updates needed?

### CODE REVIEW

File: common/defs-flags.go [hunk 1]
```diff
@@ -1,134 +1,134 @@
...
+	FlagDryRunDefinition           = "Simulate reset without resetting any Workflow Executions."
+	FlagDepthDefinition            = "Number of Child Workflows to expand, -1 to expand all Child Workflows."
+	FlagConcurrencyDefinition      = "Request concurrency."
+	FlagNoFoldDefinition           = "Disable folding. All Child Workflows within the set depth will be fetched and displayed."
 
 	// Stack trace query flag definitions
-	FlagInputSTQDefinition = "Optional Query input, in JSON format.\nFor multiple parameters, concatenate them and separate by space."
-	FlagInputFileSTQDefinition = "Passes optional Query input from a JSON file.\nIf there are multiple JSON, concatenate them and separate by space or newline.\n" + "Input from the command line will overwrite file input."
+	FlagInputSTQDefinition             = "Optional Query input, in JSON format.\nFor multiple parameters, concatenate them and separate by space."
+	FlagInputFileSTQDefinition         = "Passes optional Query input from a JSON file.\nIf there are multiple JSON, concatenate them and separate by space or newline.\n" + "Input from the command line will overwrite file input."
```
Comment (bergundy) [inline common/defs-flags.go:80]: nit: why aren't we using multiline strings here for better readability?

File: workflow/workflow_commands.go [hunk 2]
```diff
@@ -238,28 +239,32 @@ func UnmarshalMemoFromCLI(c *cli.Context) (map[string]interface{}, error) {
...
 }
 
-type historyIterator struct {
-	iter interface {
-		HasNext() bool
-		Next() (*historypb.HistoryEvent, error)
-	}
+// historyTableIter adapts history iterator for Table output view
+type historyTableIter struct {
+	iter           iterator.Iterator[*historypb.HistoryEvent]
```
Comment (bergundy) [inline workflow/workflow_commands.go:244]: What's this test failure about?

File: workflow/workflow_commands.go [hunk 3]
```diff
@@ -292,18 +297,23 @@ func printWorkflowProgress(c *cli.Context, wid, rid string, watch bool) error {
 		fmt.Println(color.Magenta(c, "Progress:"))
 	}
 
-	var lastEvent historypb.HistoryEvent // used for print result of this run
+	var lastEvent historypb.HistoryEvent
```
Comment (bergundy) [inline workflow/workflow_commands.go:300]: Why remove the comment?

File: workflow/workflow_commands.go [hunk 4]
```diff
@@ -292,18 +297,23 @@ func printWorkflowProgress(c *cli.Context, wid, rid string, watch bool) error {
...
-		Fields:     []string{"ID", "Time", "Type"},
-		FieldsLong: []string{"Details"},
-		Pager:      pager.Less,
-	}
 	errChan := make(chan error)
 	go func() {
-		hIter := sdkClient.GetWorkflowHistory(tcCtx, wid, rid, watch, enumspb.HISTORY_EVENT_FILTER_TYPE_ALL_EVENT)
-		iter := &historyIterator{iter: hIter, maxFieldLength: maxFieldLength, lastEvent: &lastEvent}
-		err = output.PrintIterator(c, iter, po)
+		iter := sdkClient.GetWorkflowHistory(tcCtx, wid, rid, watch, enumspb.HISTORY_EVENT_FILTER_TYPE_ALL_EVENT)
```
Comment (bergundy) [inline workflow/workflow_commands.go:304]: Watch and JSON aren't compatible, can we ensure they're not both set?

File: workflow/workflow_commands.go [hunk 5]
```diff
@@ -343,6 +353,25 @@ func printWorkflowProgress(c *cli.Context, wid, rid string, watch bool) error {
...
+			return err
+
+		}
+		events = append(events, event)
+	}
+
+	history := &historypb.History{}
+	history.Events = events
+
+	common.PrettyPrintJSONObject(history)
```
Comment (bergundy) [inline workflow/workflow_commands.go:370]: Just verifying that this will not generate pretty output when stdout is not a TTY.

**Review by bergundy — COMMENTED** [review]
This requires tests.

File: workflow/workflow_commands.go (hunk 3 above)
Comment (feedmeapples) [reply workflow/workflow_commands.go:300]: if `lastEvent` usage changes later in code then the comment needs to be updated as well. Figured it's better to give a good name to the variable than commenting it (the name seems to be already good)

File: workflow/workflow_commands.go (hunk 2 above)
Comment (feedmeapples) [reply workflow/workflow_commands.go:244]: requires this change https://github.com/temporalio/tctl-kit/pull/39

File: common/defs-flags.go (hunk 1 above)
Comment (feedmeapples) [reply common/defs-flags.go:80]: I'll change in a new PR

cc @DJSanti

File: workflow/workflow_commands.go (hunk 3 above)
Comment (feedmeapples) [reply workflow/workflow_commands.go]: changed to wfResult, should clarify the variable better and not need the comment?

**Review by bergundy — COMMENTED** [review]
I mostly want to see a test for this but otherwise LGTM.

**Comment by feedmeapples** [comment]
> I mostly want to see a test for this but otherwise LGTM.

added a test here separately as it requires temporaltest to exist in the server repo https://github.com/temporalio/cli/pull/171 (and to not block the current PR). One more https://github.com/temporalio/cli/pull/170

**Comment by bergundy** [comment]
I’d rather use the cli for the test server so we utilize it as much as possible

**Comment by feedmeapples** [comment]
> I’d rather use the cli for the test server so we utilize it as much as possible

i'm split between these options:

Ideally i would like integration tests to directly cover CLI dev server, the commands and options. To test one thing at a time instead of implicitly proxy testing CLI server in unrelated tests

File: workflow/workflow_commands.go [hunk 6]
```diff
@@ -343,6 +353,25 @@ func printWorkflowProgress(c *cli.Context, wid, rid string, watch bool) error {
...
+		event, err := iter.Next()
+		if err != nil {
+			return err
+
+		}
+		events = append(events, event)
+	}
+
+	history := &historypb.History{}
+	history.Events = events
```
Comment (alexshtin) [inline workflow/workflow_commands.go:368]:
```suggestion
	history := &historypb.History{Events: events}
```

File: common/util.go [hunk 7]
```diff
@@ -179,12 +179,15 @@ func PrettyPrintJSONObject(o interface{}) {
 		b, err = json.MarshalIndent(o, "", "  ")
 	}
 
+	w := c.App.Writer
+
 	if err != nil {
-		fmt.Printf("Error when try to print pretty: %v", err)
-		fmt.Println(o)
+		fmt.Fprintf(w, "Error when try to print pretty: %v", err)
```
Comment (alexshtin) [inline common/util.go:185]: All `fmt.Printf` should be `fmt.Fprintf`. No?

**Review by alexshtin — APPROVED** [review]

File: common/util.go (hunk 7 above)
Comment (feedmeapples) [reply common/util.go:185]: yes! i'll send separately to not mix

**Pull request merged**